amm/contracts/router/lib.rs

```diff
@@ -178,7 +243,8 @@ pub mod router {
             amounts[path.len() - 1] = amount_out;
             for i in (0..path.len() - 1).rev() {
                 let (reserve_0, reserve_1) = self.get_reserves(path[i], path[i + 1])?;
-                amounts[i] = self.get_amount_in(amounts[i + 1], reserve_0, reserve_1)?;
+                let fee = self.get_fee(path[i], path[i + 1])?;
```

Comment:
I bet this isn't exactly free either - instead of _knowing_ the fee we need to read it from the `Mapping`.